### PR TITLE
WIP Accelerate enabled open source models

### PR DESCRIPTION
**Pull request by thepok**
Many TODOs but already workable.

Now its easy to load big open source models.

I was able to load neox 20b and facebook 30b on my laptop with 64gb ram and 16gb vram.

### CODE REVIEW

**Review by hwchase17 — COMMENTED** [review]
nice! im not super familiar with accerlated stuff. how generic is the get_accelerated_gpt code?

File: langchain/llms/accelerate.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        from accelerate import (
+            infer_auto_device_map,
+            init_empty_weights,
+            load_checkpoint_and_dispatch,
+        )
+        from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer
+
+        # what else models are available and compatible? TODO
+
+        weights_path = "./accelerated_model_weights/" + model_name
```
[the remark this answers is not in the document]
Comment (delip) [reply langchain/llms/accelerate.py:63]: Also, why is the weights path root hardcoded to current directory "."?

File: langchain/llms/accelerate.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        if not os.path.exists(weights_path):
```
Comment (delip) [inline langchain/llms/accelerate.py:64]: Doesn't huggingface automatically do this test when you call `from_pretrained`?

File: langchain/llms/accelerate.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            model = AutoModelForCausalLM.from_pretrained(
+                model_name, torch_dtype=torch.float16
+            )
+            os.makedirs(weights_path)
+            # here an even faster method for saving and loading is described
+            # https://www.philschmid.de/deploy-gptj-sagemaker
+            model.save_pretrained(weights_path)
+        config = AutoConfig.from_pretrained(model_name)
+
+        config.use_cache = False
```
Comment (delip) [inline langchain/llms/accelerate.py:74]: why disable HF cache use and create a new cache path manually as you are doing here?

**Review by delip — CHANGES_REQUESTED** [review]

**Comment by delip** [comment]
> nice! im not super familiar with accerlated stuff. how generic is the get_accelerated_gpt code?

@hwchase17 Since I ran into similar issue and I am familiar with this, I left a bunch of comments to @thepok in a review. 

@thepok Besides the review comments, can you add an example of how Accelarate is used (preferably not as a notebook for review purposes)?

**Comment by thepok** [comment]
@delip i have no interest at the moment to implement this further. Feel free to implement your suggestions or it could be possible to merge it in the huggingface llm loader

**Comment by dev2049** [comment]
stale

**Comment by Bec-k** [comment]
ugh, so this have not been integrated?

**Comment by thepok** [comment]
nope, but i think huggingfaces transformer lib can it now by itself, and hugginfaces transformer is already in langchain :)

**Pull request closed without merging**